# Metadata Wrangling Tutorial

## Import packages

Pandas is the main package we'll use, but we'll need Numpy once or twice.

In [1]:
import pandas as pd
import numpy as np

## Load metadata Excel file

Pandas lets us import spreadsheets in several formats, including CSV, TSV, and XLS -- with `xlrd` installed (but not XLSX with the current version). Here we will import a file in XLS format.

In [2]:
df = pd.read_excel('metadata1.xls')

## Inspect metadata

### Preview

Look at the DataFrame you just imported. For a large table you will see just the first 5 and last 5 rows, and just the first 10 and last 10 columns. The total number of rows and columns is printed below the table preview.

In [3]:
df

sample_name plate_no well_id  well_no  study_no  \
0                               Vibrio1      P1a      A1        1         0   
1                             Negative1      P1a      B1        2         0   
2    Sandin54.Blanton.host.associated.6      P1a      C1        3        54   
3                    Mayer33.sediment.2      P1a      D1        4        33   
4                         Bowen74.sed15      P1a      E1        5        74   
..                                  ...      ...     ...      ...       ...   
571  Rohwer84.arctic.glacier.soil.FJLL2       P5     D12       92        84   
572                        Song51.24754       P5     E12       93        51   
573           Tucker58.DOE.coal.AP1.231       P5     F12       94        58   
574                 Angenent65.misc.519       P5     G12       95        65   
575                Berry2.sediment.Un71       P5     H12       96         2   

    principal_investigator mass_mg  plated_date plating  dna_date  ...  \
0                     none     NaN        60816     JJM     60816  ...   
1                     none     NaN        60816     JJM     60816  ...   
2                   Sandin    swab        60816     JJM     60816  ...   
3                    Mayer    swab        60816     JJM     60816  ...   
4                    Bowen    swab        60816     JJM     60816  ...   
..                     ...     ...          ...     ...       ...  ...   
571                 Rohwer     170        63016     Jon     70116  ...   
572                   Song     159        63016     Jon     70116  ...   
573                 Tucker     242        63016     Jon     70116  ...   
574               Angenent     100        63016     Jon     70116  ...   
575                  Berry      65        63016     Jon     70116  ...   

    dna_conc_cv pcr_date   primer_name primer_plate pcr_person  pcr_robot  \
0        32.066    61516  EMP515f806rB            3         KS     carmen   
1        23.026    61516  EMP515f806rB            3         KS     carmen   
2        22.957    61516  EMP515f806rB            3         KS     carmen   
3        38.331    61516  EMP515f806rB            3         KS     carmen   
4        37.511    61516  EMP515f806rB            3         KS     carmen   
..          ...      ...           ...          ...        ...        ...   
571      58.917    70516  EMP515f806rB            5        JJM       RIKE   
572      49.738    70516  EMP515f806rB            5        JJM       RIKE   
573      65.706    70516  EMP515f806rB            5        JJM       RIKE   
574      79.423    70516  EMP515f806rB            5        JJM       RIKE   
575      43.631    70516  EMP515f806rB            5        JJM       RIKE   

    pcr_primer_plate pcr_band_bool  pcr_conc pcr_ul_pooled  
0                  3             1      25.0            10  
1                  3             0       8.0            20  
2                  3             0       9.0            20  
3                  3             1      20.0            12  
4                  3             1      13.0            19  
..               ...           ...       ...           ...  
571                5             0       3.5            20  
572                5             1      22.4            11  
573                5             0       4.8            20  
574                5             1      29.7             8  
575                5             1      19.5            12  

[576 rows x 27 columns]

### Shape

Another way to see the dimensions (rows, columns) of the table is the `shape` property.

In [4]:
df.shape

(576, 27)

### Columns

List all of the column names so you can see which metadata are provided. Notice that the column names are formatted in lowercase_with_underscores (good practice). An alternative format is camelCase. It's generally good to avoid spaces and punctuation in column headers.

In [5]:
df.columns

Index(['sample_name', 'plate_no', 'well_id', 'well_no', 'study_no',
       'principal_investigator', 'mass_mg', 'plated_date', 'plating',
       'dna_date', 'dna_person', 'dna_kit_name', 'extraction_kit_lot',
       'extraction_robot', 'dna_yield_method', 'dna_conc_mean',
       'dna_conc_stdev', 'dna_conc_cv', 'pcr_date', 'primer_name',
       'primer_plate', 'pcr_person', 'pcr_robot', 'pcr_primer_plate',
       'pcr_band_bool', 'pcr_conc', 'pcr_ul_pooled'],
      dtype='object')

### Value counts

Inspect the values in some of the columns using the fuction `value_counts`.

Let's start with the column "plate_no":

In [6]:
pd.value_counts(df.plate_no)

P4     96
P1b    96
P5     96
P3     96
P1a    96
P2     96
Name: plate_no, dtype: int64

We can see there are 6 different values for 'plate_no' and each one is found 96 times. These are probably 96-well plates, and the samples are distributed over those 6 96-well plates. That's why we have 576 samples.

Next we look at the column "principal_investigator":

In [7]:
pd.value_counts(df.principal_investigator)

Song          65
none          61
Thomas        56
Rohwer        45
Zaneveld      25
Palenik       23
Mayer         23
Jensen        23
Berry         23
Metcalf       20
Bowen         19
Distel        18
Bittleston    18
Sandin        18
Kshtrika      17
Tucker        16
Uren          15
Girguis       15
King          15
Tait          13
McMahon       12
Smith         11
Minich         9
Stewart        8
Angenent       8
Name: principal_investigator, dtype: int64

These must be the names of the PIs corresponding to the samples. Song has the most samples (65), and Stewart and Angenent have the fewest samples (8). A large number of samples (61) don't have a PI. Let's learn more about those.

## Slicing a DataFrame

Panda allows you to slice a DataFrame by the values in a single column. Here we will make a new DataFrame that includes just the rows of `df` that have 'none' as the value for 'principal_investigator'.

In [8]:
df_no_pi = df[df.principal_investigator == 'none']
df_no_pi

sample_name plate_no well_id  well_no  study_no principal_investigator  \
0       Vibrio1      P1a      A1        1         0                   none   
1     Negative1      P1a      B1        2         0                   none   
6     Negative2      P1a      G1        7         0                   none   
7       Vibrio2      P1a      H1        8         0                   none   
8     Negative3      P1a      A2        9         0                   none   
..          ...      ...     ...      ...       ...                    ...   
540   Negative4       P5      E8       61         0                   none   
548   Negative5       P5      E9       69         0                   none   
549   Negative6       P5      F9       70         0                   none   
557   Negative7       P5     F10       78         0                   none   
570   Negative8       P5     C12       91         0                   none   

    mass_mg  plated_date plating  dna_date  ... dna_conc_cv pcr_date  \
0       NaN        60816     JJM     60816  ...      32.066    61516   
1       NaN        60816     JJM     60816  ...      23.026    61516   
6       NaN        60816     JJM     60816  ...      20.159    61516   
7       NaN        60816     JJM     60816  ...      35.474    61516   
8       NaN        60816     JJM     60816  ...      26.357    61516   
..      ...          ...     ...       ...  ...         ...      ...   
540       0        63016     Jon     70116  ...      17.949    70516   
548       1        63016     Jon     70116  ...      26.819    70516   
549       0        63016     Jon     70116  ...      15.556    70516   
557       0        63016     Jon     70116  ...      54.788    70516   
570       0        63016     Jon     70116  ...      67.793    70516   

      primer_name primer_plate pcr_person  pcr_robot pcr_primer_plate  \
0    EMP515f806rB            3         KS     carmen                3   
1    EMP515f806rB            3         KS     carmen                3   
6    EMP515f806rB            3         KS     carmen                3   
7    EMP515f806rB            3         KS     carmen                3   
8    EMP515f806rB            3         KS     carmen                3   
..            ...          ...        ...        ...              ...   
540  EMP515f806rB            5        JJM       RIKE                5   
548  EMP515f806rB            5        JJM       RIKE                5   
549  EMP515f806rB            5        JJM       RIKE                5   
557  EMP515f806rB            5        JJM       RIKE                5   
570  EMP515f806rB            5        JJM       RIKE                5   

    pcr_band_bool  pcr_conc pcr_ul_pooled  
0               1      25.0            10  
1               0       8.0            20  
6               0      12.0            19  
7               1      23.0            10  
8               0      16.0            15  
..            ...       ...           ...  
540             0       7.6            20  
548             0       7.4            20  
549             0       5.1            20  
557             0       3.5            20  
570             0      20.0            12  

[61 rows x 27 columns]

As expected, this column has 61 rows (and 27 columns).

In [9]:
pd.value_counts(df_no_pi.study_no)

0    61
Name: study_no, dtype: int64

We can see that all of the 'study_no' values are 0. This must not be a normal kind of study.

In [10]:
df_no_pi['sample_name']

0        Vibrio1
1      Negative1
6      Negative2
7        Vibrio2
8      Negative3
         ...    
540    Negative4
548    Negative5
549    Negative6
557    Negative7
570    Negative8
Name: sample_name, Length: 61, dtype: object

Looking at the sample names from the first column, 'sample_name', we can see that they all have the word 'Vibrio', 'Negative', or 'control' in them. These could be the negative and positive controls, which would explain why they don't have a normal study number and PI name. 

In [11]:
pd.value_counts(df_no_pi['sample_name'])

Negative4                                     6
Negative2                                     6
Negative3                                     6
Negative6                                     6
Negative1                                     6
Negative7                                     6
Negative5                                     6
Negative8                                     3
Vibrio3                                       2
Negative9                                     2
Vibrio2                                       2
control.soil.under.pine.tree.by.Skaggs        2
control.soil.under.rock.near.century.plant    2
Vibrio1                                       2
Vibrio4                                       2
control.soil.grass.near.BRF                   2
Name: sample_name, dtype: int64

We can also see that some of the sample names are duplicated.

## Refine Metadata File #1

### Dates

We can see that the two columns with dates are formatted strangely.

In [12]:
df[['plated_date', 'dna_date']]

plated_date  dna_date
0          60816     60816
1          60816     60816
2          60816     60816
3          60816     60816
4          60816     60816
..           ...       ...
571        63016     70116
572        63016     70116
573        63016     70116
574        63016     70116
575        63016     70116

[576 rows x 2 columns]

Let's assume we know that these are in the format 'MDDYY', and we want to put them into Pandas datatime format. We can store the first date in the first column to work with as an example.

In [13]:
date_int = df.loc[0, 'plated_date']
date_int

60816

The `type` function shows that the date value is stored as an integer.

In [14]:
type(date_int)

numpy.int64

In [15]:
date_str = str(date_int)
date_str

'60816'

We can pull out the parts of the date using substrings. In Python, strings are stored as arrays of characters, so we can use array-style indexing to access the substrings.

In [16]:
# month
month = int(date_str[0])
month

6

In [17]:
# day
day = int(date_str[1:3])
day

8

In [18]:
# year
year = int(date_str[3:5]) + 2000
year

2016

In [19]:
date_pd = pd.datetime(year, month, day)
date_pd

datetime.datetime(2016, 6, 8, 0, 0)

In [20]:
print(date_pd)

2016-06-08 00:00:00


We can write a function to do the above for all of the dates.

In [21]:
def convert_date(date):
    date = str(date)
    month = int(date[0])
    day = int(date[1:3])
    year = int(date[3:5]) + 2000
    return(pd.datetime(year, month, day))

In [22]:
convert_date(61220)

datetime.datetime(2020, 6, 12, 0, 0)

Finally, we can apply this function to the DataFrame to create new columns for latitude and longitude. We will use what's called a list comprehension in Python.

In [23]:
df['plated_date_fixed'] = [convert_date(x) for x in df.plated_date]

In [24]:
df['dna_date_fixed'] = [convert_date(x) for x in df.dna_date]

We can look at the original two date columns and the two we created from them.

In [25]:
df[['plated_date', 'plated_date_fixed', 'dna_date', 'dna_date_fixed']]

plated_date plated_date_fixed  dna_date dna_date_fixed
0          60816        2016-06-08     60816     2016-06-08
1          60816        2016-06-08     60816     2016-06-08
2          60816        2016-06-08     60816     2016-06-08
3          60816        2016-06-08     60816     2016-06-08
4          60816        2016-06-08     60816     2016-06-08
..           ...               ...       ...            ...
571        63016        2016-06-30     70116     2016-07-01
572        63016        2016-06-30     70116     2016-07-01
573        63016        2016-06-30     70116     2016-07-01
574        63016        2016-06-30     70116     2016-07-01
575        63016        2016-06-30     70116     2016-07-01

[576 rows x 4 columns]

## Refine Metadata File #1

For the next few exercises we will use a second metadata file.

In [26]:
df2 = pd.read_excel('metadata2.xls')

In [27]:
df2

DNA-ID (unique)   Name_on_tube Date_Sample_(DDMonYYY) Sample_Source  \
0            D0050              1             2017-06-03         Blank   
1            D0051  122_425_100_1             2018-06-03    CTD_Lasker   
2            D0052  122_425_100_3             2018-06-03    CTD_Lasker   
3            D0053  122_425_100_5             2018-06-03    CTD_Lasker   
4            D0054  122_425_100_7             2018-06-03    CTD_Lasker   
5            D0055  122_425_100_9             2018-06-03    CTD_Lasker   
6            D0056  122_425_18_15             2018-06-03    CTD_Lasker   
7            D0057  122_425_18_17             2018-06-03    CTD_Lasker   
8            D0058  122_425_18_19             2018-06-03    CTD_Lasker   
9            D0059  122_425_18_21             2018-06-03    CTD_Lasker   
10           D0060  122_425_18_23             2018-06-03    CTD_Lasker   
11           D0061             12             2018-06-04         Blank   
12           D0062  123_422_100_1             2018-06-04    CTD_Lasker   
13           D0063  123_422_100_3             2018-06-04    CTD_Lasker   
14           D0064  123_422_100_5             2018-06-04    CTD_Lasker   
15           D0065  123_422_100_7             2018-06-04    CTD_Lasker   
16           D0066  123_422_100_9             2018-06-04    CTD_Lasker   
17           D0067  123_422_10_15             2018-06-04    CTD_Lasker   
18           D0068  123_422_10_17             2018-06-04    CTD_Lasker   
19           D0069  123_422_10_19             2018-06-04    CTD_Lasker   
20           D0070  123_422_10_21             2018-06-04    CTD_Lasker   
21           D0071  123_422_10_23             2018-06-04    CTD_Lasker   

   Time_Sample_(local24hr)         DateTimeGMT       lat       long  \
0                      NaN                 NaT       NaN        NaN   
1                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
2                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
3                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
4                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
5                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
6                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
7                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
8                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
9                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
10                20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
11                     NaN                 NaT       NaN        NaN   
12                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
13                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
14                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
15                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
16                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
17                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
18                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
19                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
20                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
21                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   

                            Location Station  CTD Nisken#  SampleDepth(m)  
0   blank-MQ-fromMBARI-notautoclaved       B    B       B             NaN  
1                  San Miguel (dusk)     425  122       1           100.0  
2                  San Miguel (dusk)     425  122       3           100.0  
3                  San Miguel (dusk)     425  122       5           100.0  
4                  San Miguel (dusk)     425  122       7           100.0  
5                  San Miguel (dusk)     425  122       9           100.0  
6                  San Miguel (dusk)     425  122      15            18.0  
7                  San Miguel (dusk)     425  122      17 

### Datetimes

Notice there is a column called "DateTimeGMT", which appears to have been automatically imported with each value as a datetime object. Let's confirm this. Viewing the column, we see at the bottom that the datatype (dtype) is "datetime64[ns]".

In [28]:
df2['DateTimeGMT']

0                    NaT
1    2018-06-03 03:58:00
2    2018-06-03 03:58:00
3    2018-06-03 03:58:00
4    2018-06-03 03:58:00
5    2018-06-03 03:58:00
6    2018-06-03 03:58:00
7    2018-06-03 03:58:00
8    2018-06-03 03:58:00
9    2018-06-03 03:58:00
10   2018-06-03 03:58:00
11                   NaT
12   2018-06-04 13:50:00
13   2018-06-04 13:50:00
14   2018-06-04 13:50:00
15   2018-06-04 13:50:00
16   2018-06-04 13:50:00
17   2018-06-04 13:50:00
18   2018-06-04 13:50:00
19   2018-06-04 13:50:00
20   2018-06-04 13:50:00
21   2018-06-04 13:50:00
Name: DateTimeGMT, dtype: datetime64[ns]

Let's create a new column that contains just the date.

The `date()` function give us this power.

In [29]:
df2.loc[1, 'DateTimeGMT'].date()

datetime.date(2018, 6, 3)

Now we can add that new column.

In [30]:
df2['DateTimeGMT_date'] = [x.date() for x in df2.DateTimeGMT]

Let's check what we added.

In [31]:
df2[['DateTimeGMT', 'DateTimeGMT_date']]

DateTimeGMT DateTimeGMT_date
0                  NaT              NaT
1  2018-06-03 03:58:00       2018-06-03
2  2018-06-03 03:58:00       2018-06-03
3  2018-06-03 03:58:00       2018-06-03
4  2018-06-03 03:58:00       2018-06-03
5  2018-06-03 03:58:00       2018-06-03
6  2018-06-03 03:58:00       2018-06-03
7  2018-06-03 03:58:00       2018-06-03
8  2018-06-03 03:58:00       2018-06-03
9  2018-06-03 03:58:00       2018-06-03
10 2018-06-03 03:58:00       2018-06-03
11                 NaT              NaT
12 2018-06-04 13:50:00       2018-06-04
13 2018-06-04 13:50:00       2018-06-04
14 2018-06-04 13:50:00       2018-06-04
15 2018-06-04 13:50:00       2018-06-04
16 2018-06-04 13:50:00       2018-06-04
17 2018-06-04 13:50:00       2018-06-04
18 2018-06-04 13:50:00       2018-06-04
19 2018-06-04 13:50:00       2018-06-04
20 2018-06-04 13:50:00       2018-06-04
21 2018-06-04 13:50:00       2018-06-04

### Missing values

Let's say we want to filter out rows having missing values for some columns, for example, "SampleDepth(m)". We can do this with the Pandas function `isna()`.

In [32]:
df2[df2['SampleDepth(m)'].isna()]

DNA-ID (unique) Name_on_tube Date_Sample_(DDMonYYY) Sample_Source  \
0            D0050            1             2017-06-03         Blank   
11           D0061           12             2018-06-04         Blank   

   Time_Sample_(local24hr) DateTimeGMT  lat long  \
0                      NaN         NaT  NaN  NaN   
11                     NaN         NaT  NaN  NaN   

                            Location Station CTD Nisken#  SampleDepth(m)  \
0   blank-MQ-fromMBARI-notautoclaved       B   B       B             NaN   
11  blank-MQ-fromMBARI-notautoclaved       B   B       B             NaN   

   DateTimeGMT_date  
0               NaT  
11              NaT

To get the columns that are *not* "NaN", we can precede the operation with `~` (not).

In [33]:
df2_filt = df2[~df2['SampleDepth(m)'].isna()]

In [34]:
df2_filt

DNA-ID (unique)   Name_on_tube Date_Sample_(DDMonYYY) Sample_Source  \
1            D0051  122_425_100_1             2018-06-03    CTD_Lasker   
2            D0052  122_425_100_3             2018-06-03    CTD_Lasker   
3            D0053  122_425_100_5             2018-06-03    CTD_Lasker   
4            D0054  122_425_100_7             2018-06-03    CTD_Lasker   
5            D0055  122_425_100_9             2018-06-03    CTD_Lasker   
6            D0056  122_425_18_15             2018-06-03    CTD_Lasker   
7            D0057  122_425_18_17             2018-06-03    CTD_Lasker   
8            D0058  122_425_18_19             2018-06-03    CTD_Lasker   
9            D0059  122_425_18_21             2018-06-03    CTD_Lasker   
10           D0060  122_425_18_23             2018-06-03    CTD_Lasker   
12           D0062  123_422_100_1             2018-06-04    CTD_Lasker   
13           D0063  123_422_100_3             2018-06-04    CTD_Lasker   
14           D0064  123_422_100_5             2018-06-04    CTD_Lasker   
15           D0065  123_422_100_7             2018-06-04    CTD_Lasker   
16           D0066  123_422_100_9             2018-06-04    CTD_Lasker   
17           D0067  123_422_10_15             2018-06-04    CTD_Lasker   
18           D0068  123_422_10_17             2018-06-04    CTD_Lasker   
19           D0069  123_422_10_19             2018-06-04    CTD_Lasker   
20           D0070  123_422_10_21             2018-06-04    CTD_Lasker   
21           D0071  123_422_10_23             2018-06-04    CTD_Lasker   

   Time_Sample_(local24hr)         DateTimeGMT       lat       long  \
1                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
2                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
3                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
4                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
5                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
6                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
7                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
8                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
9                 20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
10                20:58:00 2018-06-03 03:58:00  33 54.51  120 41.96   
12                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
13                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
14                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
15                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
16                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
17                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
18                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
19                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
20                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   
21                06:50:00 2018-06-04 13:50:00  34 24.36  120 21.67   

             Location Station  CTD Nisken#  SampleDepth(m) DateTimeGMT_date  
1   San Miguel (dusk)     425  122       1           100.0       2018-06-03  
2   San Miguel (dusk)     425  122       3           100.0       2018-06-03  
3   San Miguel (dusk)     425  122       5           100.0       2018-06-03  
4   San Miguel (dusk)     425  122       7           100.0       2018-06-03  
5   San Miguel (dusk)     425  122       9           100.0       2018-06-03  
6   San Miguel (dusk)     425  122      15            18.0       2018-06-03  
7   San Miguel (dusk)     425  122      17            18.0       2018-06-03  
8   San Miguel (dusk)     425  122      19            18.0       2018-06-03  
9   San Miguel (dusk)     425  122      21            18.0       2018-06-03  
10  San Miguel (dusk)     425  122      23            18.0       2018-06-03  
12  San Miguel (dawn)     422  123       1           100.0       2018-06-04  
13  San Miguel (dawn

### Latitude and longitude

We can see that the latitude and longitude columns are degrees and minutes separated by a space, which is not immediately machine readable. It might be better to have them as decimal-degrees.

In [35]:
df2[['lat', 'long']]

lat       long
0        NaN        NaN
1   33 54.51  120 41.96
2   33 54.51  120 41.96
3   33 54.51  120 41.96
4   33 54.51  120 41.96
5   33 54.51  120 41.96
6   33 54.51  120 41.96
7   33 54.51  120 41.96
8   33 54.51  120 41.96
9   33 54.51  120 41.96
10  33 54.51  120 41.96
11       NaN        NaN
12  34 24.36  120 21.67
13  34 24.36  120 21.67
14  34 24.36  120 21.67
15  34 24.36  120 21.67
16  34 24.36  120 21.67
17  34 24.36  120 21.67
18  34 24.36  120 21.67
19  34 24.36  120 21.67
20  34 24.36  120 21.67
21  34 24.36  120 21.67

Let's make a function that coverts these to latitude and longitude. We add an if-statement to check if the value we are converting is "nan" (not a number).

In [36]:
def convert_lat_lon(x):
    s = str(x)
    if s != 'nan':        
        degrees, minutes = s.split(' ')
        value = float(degrees) + float(minutes)/60
        return(value)
    else:
        return(np.nan)

In [37]:
df2.loc[1, 'lat']

'33 54.51'

In [38]:
convert_lat_lon(df2.loc[1, 'lat'])

33.9085

Now that we are happy the function works, we can apply it to the DataFrame.

In [39]:
df2['latitude'] = [convert_lat_lon(x) for x in df2.lat]

In [40]:
df2['longitude'] = [convert_lat_lon(x) for x in df2.long]

We can check the columns we added and the ones we started with.

In [41]:
df2[['lat', 'latitude', 'long', 'longitude']]

lat  latitude       long   longitude
0        NaN       NaN        NaN         NaN
1   33 54.51   33.9085  120 41.96  120.699333
2   33 54.51   33.9085  120 41.96  120.699333
3   33 54.51   33.9085  120 41.96  120.699333
4   33 54.51   33.9085  120 41.96  120.699333
5   33 54.51   33.9085  120 41.96  120.699333
6   33 54.51   33.9085  120 41.96  120.699333
7   33 54.51   33.9085  120 41.96  120.699333
8   33 54.51   33.9085  120 41.96  120.699333
9   33 54.51   33.9085  120 41.96  120.699333
10  33 54.51   33.9085  120 41.96  120.699333
11       NaN       NaN        NaN         NaN
12  34 24.36   34.4060  120 21.67  120.361167
13  34 24.36   34.4060  120 21.67  120.361167
14  34 24.36   34.4060  120 21.67  120.361167
15  34 24.36   34.4060  120 21.67  120.361167
16  34 24.36   34.4060  120 21.67  120.361167
17  34 24.36   34.4060  120 21.67  120.361167
18  34 24.36   34.4060  120 21.67  120.361167
19  34 24.36   34.4060  120 21.67  120.361167
20  34 24.36   34.4060  120 21.67  120.361167
21  34 24.36   34.4060  120 21.67  120.361167

When we use decimal-degrees for latitude and longitude, we need to be mindful of the sign (positive or negative). That is, latitude values north of the equator should be positive and south of the equator should be negative; longitude values east of the prime meridian should be positive and west of the prime meridian should be negative.

Let's spot-check one of these values. If we seach "34.4060, 120.361167" in Google Maps, we see this coordinate pair is in the Yellow Sea off the coast of China. Probably the longitude values should be negative! Yes, if we instead search for "34.4060, -120.361167", it's a point off the coast of Southern California, as expected.

So let's change that last command to negate the longitude values for this dataset.

In [42]:
df2['longitude'] = [convert_lat_lon(x)*-1 for x in df2.long]

In [43]:
df2[['lat', 'latitude', 'long', 'longitude']].head()

lat  latitude       long   longitude
0       NaN       NaN        NaN         NaN
1  33 54.51   33.9085  120 41.96 -120.699333
2  33 54.51   33.9085  120 41.96 -120.699333
3  33 54.51   33.9085  120 41.96 -120.699333
4  33 54.51   33.9085  120 41.96 -120.699333